google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1811,6 +1813,113 @@ Blob create(
   @Deprecated
   Blob create(BlobInfo blobInfo, InputStream content, BlobWriteOption... options);
 
+  /**
+   * Uploads {@code path} to the blob using {@link #writer}. By default any MD5 and CRC32C values in
+   * the given {@code blobInfo} are ignored unless requested via the {@link
+   * BlobWriteOption#md5Match()} and {@link BlobWriteOption#crc32cMatch()} options. Folder upload is
+   * not supported.
+   *
+   * <p>Example of uploading a file:
+   *
+   * <pre>{@code
+   * String bucketName = "my-unique-bucket";
+   * String fileName = "readme.txt";
+   * BlobId blobId = BlobId.of(bucketName, fileName);
+   * BlobInfo blobInfo = BlobInfo.newBuilder(blobId).setContentType("text/plain").build();
+   * storage.upload(blobInfo, Paths.get(fileName));
+   * }</pre>
+   *
+   * @param blobInfo blob to create
+   * @param path file to upload
+   * @param options blob write options
+   * @throws IOException on I/O error
+   * @throws StorageException on failure
+   * @see #upload(BlobInfo, Path, int, BlobWriteOption...)
+   */
+  void upload(BlobInfo blobInfo, Path path, BlobWriteOption... options) throws IOException;
```

Comment:
Blob should be returned from upload() methods similar to create() methods.

Reply:
returning Blob requires an extra RPC request to be issued, that is not always needed.
I can add uploadFrom() method to the Blob class (symmetric to downloadTo)  to return an updated Blob.

Reply:
A month ago I compared upload/download implemented in various ways for several types of file. This is the [summary](https://docs.google.com/document/d/19u1wYn-8kDdz9bv2m106ApGmZV3KF8__EGuQSn1g2gU/edit#)

Reply:
Do you mean we can update the StorageRpc interface to make void write(uploadId, ...) to return a StorageObject?

Reply:
> 1 nit so far,  added @JesseLovelace for surface review as well.
> Blob should be returned from upload() methods similar to create() methods.

It's not a nit at all :) It's a significant change. I agree with you, upload() should return Blob.
I see the following ways how to implement:

1) Update the StorageRpc interface and make write() to return a StorageObject instance or null in  write is not final request for the given upload. Now StorageRpc .write() returns null.

2) Update HttpStorageRpc.write() to remember StorageObject  and return this object in a separate method.
In this case BlobWriterChannel.fluchBuffer() will look like:
```
public void run() {
         StorageRpc  rpc = getOptions().getStorageRpcV1();
         rpc.write(getUploadId(), getBuffer(), 0, getPosition(), length, last);
         if (last && rpc instnceof HttpStorageRpc) {
             this.uploadedBlob = ((HttpStorageRpc)rpc.getUplodedBlob())
         } else {
             this.uploadedBlob = null;  // let storage.upload() to resolve it
         }
}
```
3) Issue an extra rpc per upload to return a Blob

I believe 2. is not a Google way, so we should go 1. As a temporary solution 3. is also considered, because it could be improved to 1.

WDYT?

Reply:
Hi @frankyn,

> When the upload is complete GCS responds with object metadata so we don't need to perform an additional get request.

I managed to implement upload based on your hint. I modified `StorageRpc.write()` to return `StorageObject` instead of `void`.
Two issues were found:
- GCS does not respond with object metadata if writer was created by `storage.writer(signedUrl)`
Not sure if it's a bug or a feature, looks like a bug: [spec](https://developers.google.com/gdata/docs/resumable_upload#Upload) says nothing about it. I worked it around.

- Change StrorageRpc affects storage-nio, Kokoro - Test: Linkage Monitor fails:
`(com.google.cloud:google-cloud-nio:0.120.0-alpha) com.google.cloud.storage.contrib.nio.testing.FakeStorageRpc's method write(String arg1, byte[] arg2, int arg3, long arg4, int arg6, boolean arg7) is not implemented in the class referenced from com.google.cloud.storage.spi.v1.StorageRpc (com.google.cloud:google-cloud-storage:1.107.1-SNAPSHOT)
`

cc: @JesseLovelace